src/main/java/org/dsa/iot/alarm/AlarmAlgorithm.java

```diff
@@ -0,0 +1,362 @@
+ * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
+ * IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+package org.dsa.iot.alarm;
+
+import org.dsa.iot.dslink.node.*;
+import org.dsa.iot.dslink.node.actions.*;
+import org.dsa.iot.dslink.node.value.*;
+import org.dsa.iot.dslink.util.*;
+import org.dsa.iot.dslink.util.json.*;
+import java.util.concurrent.*;
+
+/**
+ * Alarm algorithms evaluate the state of AlarmWatch objects, and generate alarms for
+ * each when the conditions of the algorithm are met.
+ * <p/>
+ * Subclasses should override the following methods:
+ * <ul>
+ * <li>isAlarm(AlarmWatch).</li>
+ * <li>getAlarmMessage(AlarmWatch).</li>
+ * </ul>
+ *
+ * @author Aaron Hansen
+ */
+public abstract class AlarmAlgorithm extends AbstractAlarmObject implements Runnable {
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constants
+    ///////////////////////////////////////////////////////////////////////////
+
+    private static final String ALARM_TYPE = "Alarm Type";
+    private static final String AUTO_UPDATE_INTERVAL = "Auto Update Interval";
+    private static final String TO_ALARM_INHIBIT = "To Alarm Inhibit";
+    private static final String TO_NORMAL_INHIBIT = "To Normal Inhibit";
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    private AlarmState alarmState; //alert, fault or offnormal
+    private ScheduledFuture autoUpdateFuture;
+    private boolean updatingAll = false;
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ///////////////////////////////////////////////////////////////////////////
+
+    public AlarmAlgorithm() {
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Methods
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Action handler for adding watches for alarm sources.
+     */
+    protected void addWatch(final ActionResult event) {
+        try {
+            Value name = event.getParameter(NAME);
+            if (name == null) {
+                throw new IllegalArgumentException("Missing name");
+            }
+            String nameString = name.getString();
+            Node parent = event.getNode().getParent();
+            Node watchNode = parent.getChild(nameString);
+            if (watchNode != null) {
+                throw new IllegalArgumentException(
+                        "Name already in use: " + name.getString());
+            }
+            Value path = event.getParameter(PATH);
+            if (path == null) {
+                throw new NullPointerException("Missing path");
```

Comment:
Earlier in the PR, I saw a NPE with just `uuid` as the message. You might want to keep the same convention throughout your code.